internal/report/report.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2020 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+// Package report provides functions to report OPA's version information to an external service and process the response.
+package report
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"os"
+	"strings"
+	"time"
+
+	"github.com/open-policy-agent/opa/plugins/rest"
+	"github.com/open-policy-agent/opa/util"
+	"github.com/open-policy-agent/opa/version"
+)
+
+const (
+	defaultExternalServiceURL = "https://telemetry.openpolicyagent.org"
+)
+
+// Reporter reports the version of the running OPA instance to an external service
+type Reporter struct {
+	body   map[string]string
+	client rest.Client
+}
+
+// DataResponse represents the data returned by the external service
+type DataResponse struct {
+	Latest ReleaseDetails `json:"latest,omitempty"`
+}
+
+// ReleaseDetails holds information about the latest OPA release
+type ReleaseDetails struct {
+	Download     string `json:"download,omitempty"`      // link to download the OPA release
+	ReleaseNotes string `json:"release_notes,omitempty"` // link to the OPA release notes
+}
+
+// New returns an instance of the Reporter
+func New(id string) (*Reporter, error) {
+	r := Reporter{}
+	r.body = map[string]string{
+		"id":      id,
+		"version": version.Version,
+	}
+
+	url := os.Getenv("OPA_TELEMETRY_SERVICE_URL")
+	if url == "" {
+		url = defaultExternalServiceURL
+	}
+
+	restConfig := []byte(fmt.Sprintf(`{
+		"url": %q,
+	}`, url))
+
+	client, err := rest.New(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	r.client = client
+
+	return &r, nil
+}
+
+// SendReport sends the version report to the external service
+func (r *Reporter) SendReport(ctx context.Context) (*DataResponse, error) {
+	rCtx, cancel := context.WithTimeout(ctx, 1*time.Second)
+	defer cancel()
+
+	resp, err := r.client.WithJSON(r.body).Do(rCtx, "POST", "/v1/version")
```

Comment:
This is not per the document in this PR: this delays the OPA startup for the duration of the API call.